expression/builtin_cast.go

```diff
@@ -76,6 +78,192 @@ func (b *builtinCastSig) eval(row []types.Datum) (d types.Datum, err error) {
 	return d, errors.Errorf("unknown cast type - %v", b.tp)
 }
 
+type castAsIntFunctionClass struct {
+	baseFunctionClass
+
+	tp *types.FieldType
+}
+
+func (b *castAsIntFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	bf := baseIntBuiltinFunc{newBaseBuiltinFuncWithTp(args, b.tp, ctx)}
+	if IsHybridType(args[0]) {
+		return &builtinCastIntAsIntSig{bf}, nil
+	}
+	switch args[0].GetTypeClass() {
+	case types.ClassInt:
+		sig = &builtinCastIntAsIntSig{bf}
+	case types.ClassReal:
+		sig = &builtinCastRealAsIntSig{bf}
+	case types.ClassDecimal:
+		sig = &builtinCastDecimalAsIntSig{bf}
+	case types.ClassString:
+		switch args[0].GetType().Tp {
+		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp:
+			sig = &builtinCastTimeAsIntSig{bf}
+		case mysql.TypeDuration:
+			sig = &builtinCastDurationAsIntSig{bf}
+		default:
+			sig = &builtinCastStringAsIntSig{bf}
+		}
```

Comment:
No, `mysql.TypeDuration` is regarded as `types.ClassString`
We first check the `TypeClass` of an Expression,
if it's `types.ClassString`, then we check whether it is `mysql.TypeDuration`.